Add genetic demultiplexing component

wxicu opened the pull request:
Issues to be solved by adding genetic demultiplexing component:
1. Can we use public available containers, e.g. these provided by biocontainers, or we need to define these by our own?
2. How can I add data for the test scripts? I have used the files demuxafy_test_data, but sill need to upload more files.


Thanks!

VladimirShitov commented on src/gene_demulti/cellsnp/config.vsh.yaml:
```diff
@@ -0,0 +1,108 @@
+      description: Number of subprocesses.
+    - name: "--chrom"
+      type: string
+      description: The chromosomes to use, comma separated
```
In which format, `chr1` or `1`?

VladimirShitov commented on src/gene_demulti/freebayes/config.vsh.yaml:
```diff
@@ -0,0 +1,281 @@
+  name: freebayes
+  namespace: gene_demulti
+  description: |
+    Run freebayes
```
Could be more detailed as well

VladimirShitov commented on src/gene_demulti/souporcell/config.vsh.yaml:
```diff
@@ -0,0 +1,85 @@
+  name: souporcell
+  namespace: gene_demulti
+  description: |
+    Run Souporcell
```
Could be more detailed :)

VladimirShitov reviewed (commented):
Looks good to me :)

wxicu commented on dockerimage/popscle/Dockerfile:
```diff
@@ -0,0 +1,55 @@
+# based on the dockerfile from https://github.com/statgen/popscle
+
+# Source Image
+FROM ubuntu:18.04
```
hey, thanks for your suggestion. I have now specified the docker configuration directly in the config file. To avoid any misunderstanding, I will delete images filer under the dockerimage folder.
popscle works on the newest ubuntu, will update it.

wxicu commented on dockerimage/samtools/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+FROM python:3.9
```
will lift all up to python3.9

rcannood commented on src/gene_demulti/bcftools/config.vsh.yaml:
```diff
@@ -0,0 +1,42 @@
+  test_resources:
+    - type: bash_script
+      path: test.sh
+    - path: ../../../resources_test/demuxafy_test_data
```
Could you create a script in `workflows/resources_test_scripts/demuxafy_test_data.sh` that creates the test data for this component?

rcannood commented on src/gene_demulti/vireo/test.sh:
```diff
@@ -0,0 +1,11 @@
+echo "$meta_resources_dir"
+echo ">>> Running executable"
+$meta_executable \
+    --cellData "$meta_resources_dir/vireo_test_data/cells.cellSNP.vcf.gz" \
```
Can you create a script `workflows/resources_test_data/vireo_test_data.sh`?

rcannood reviewed (changes requested):
Hi Xichen!

Sorry for taking so long to get back to this PR.

The code looks good overall!

A few general questions:

* I'd like to ask you to add the scripts to generate the test data to the `workflows/resources_test_scripts` because otherwise the CI won't be able to run the unit tests successfully.

* Should these components be more or less interchangeable? If so, it might be a good idea to change the components such that the inputs and the outputs are all in the same format. I'd be happy to jump in a meeting to discuss this in more detail.

* Are parts of the source code based on demuxafy?

* Is it possible to start from more general images such as `python:3.10`?

DriesSchaumont commented on src/gene_demulti/scSplit/script.sh:
```diff
@@ -0,0 +1,40 @@
+  mkdir $par_output
+fi
+
+scSplit_loc="/usr/local/lib/python3.9/site-packages/scSplit"
```
`scSplit` has a really unconventional way to install. I would really like this path to be python version agnostic. Could you install `scSplit` using the docker `run` directive (e.g. download it using wget and copying the script to `/usr/local/bin` or `/usr/bin`)

DriesSchaumont reviewed (changes requested):
Nice PR overall! Left some small comments/suggestions

rcannood commented on src/gene_demulti/freemuxlet/config.vsh.yaml:
```diff
@@ -0,0 +1,178 @@
+    - path: ../../../resources_test/demuxafy_test_data
+platforms:
+  - type: docker
+    image: ubuntu:20.04
```
Isn't it better to pin to a specific version? So 20.04 or 22.04.

wxicu replied:
For freemuxlet/demuxlet: it works only up to 20.04. For bcftools and cellsnp it is ok

rcannood commented on src/gene_demulti/samtools/config.vsh.yaml:
```diff
@@ -0,0 +1,38 @@
+    - type: apt
+      packages: [ wget, gcc, make, libbz2-dev, zlib1g-dev, libncurses5-dev, libncursesw5-dev, liblzma-dev]
+    - type: docker
+      run: wget https://github.com/samtools/samtools/releases/download/1.16.1/samtools-1.16.1.tar.bz2 && tar jxf samtools-1.16.1.tar.bz2 && rm samtools-1.16.1.tar.bz2 && cd samtools-1.16.1 && ./configure --prefix $(pwd) && make
```
The traditional approach here is `./configure --prefix /userlocal/bin && make && make install` without changing the path. Something like this is also okay, so I wouldn't let this hold up the pr.

wxicu commented:
Hi! Thank you for the comments and hope you had a nice Christmas!

- resources_test_scripts are added
- For cellsnp and bcftools I have changed the docker configuration to ubuntu:latest, but it can be changed easily if we agree on specifying a specific ubuntu version. 
- For freemuxlet and demuxlet: It doesn't work on my Macos arm64 architecture currently to either build container from dockerfile or by openpipeline (it used to work...). I also tried [Dries configuration](https://github.com/openpipelines-bio/openpipeline/blob/feature/count_demultiplexing/src/deconvolution/freemuxlet/config.vsh.yaml). All return segmentation fault. I am still looking for a solution. But it works on other platforms...
- The source code is independent of  demuxafy. I only use some test datasets from it.
- For the input: The genetic demultiplexing tools have the bam input in common except for vireo. There are some differences in other input requirements. I will try to make a table to discuss in the meeting later.
- For the output: Although different tools have different output, but they all have a file (tsv, csv, etc) to save the genetic demultiplexing result(singlet/doublet/unassigned, and donor identity). In the nextflow pipeline I implemented before, I have a process to summarize all the genetic demultiplexing results into a single table. Is is better to have such a summarizing step at the end after running different tools or to convert to a specific format every time we run a component? But yes again, we can discuss the details in the meeting.

wxicu commented:
Hey hey, a small update:

- For the output format, each component has an output file in the same format called assignment.tsv
- For the input format: they have different input so it is difficult to let them have the input in the same format
<img width="959" alt="input_gene_demulti" src="https://user-images.githubusercontent.com/102925032/213324372-c9788120-1b1d-40ab-a44e-2f7fe84d6263.png">

rcannood commented:
Hey @wxicu! Thanks for the update :) 

* I ran the resources test script. Unfortunately, the resulting files are quite big:

  ![Screenshot from 2023-01-19 15-32-34](https://user-images.githubusercontent.com/553642/213469792-1d52b142-694c-4140-b9c3-765142e2aad8.png)

  * Is it possible to subset the bam file?
  * In addition, would you be able to use the fa/fa.fai files that are already available at `resources_test/cellranger_tiny_fastq/cellranger_tiny_ref/fasta`? These are much smaller since they are subset to contain only chr 21, and it's already on S3 so it wouldn't consume any additional storage space.

* demuxlet and freemuxlet are both missing a file called `summary.R`. Where is this R script located?

rcannood commented:
Also @wxicu, we've gone back and forth a few times now. Would you like to have a one on one meeting so we can fix the remaining issues and finalize the PR asap? :)

wxicu commented:
@rcannood Thanks for viewing the PR again!

- Actually, I think the bam file should be already on the S3, as it is also used by [demuxlet component (https://github.com/openpipelines-bio/openpipeline/blob/feature/count_demultiplexing/src/deconvolution/demuxlet/test.sh) here. 
- Souporcell didnt work with this genome reference because vatrix could not find sequence in FASTA. So I just subsetted the bam and genome by chr1 and uploaded to the dropbox. The download script has been changed correspondingly. Hope the data is smaller now!
- Oh Sorry I forgot to upload it! Just Done.

rcannood commented:
Not sure what's going on with this merge. Closing and reopening.
